File: yarl-java/src/test/java/hr/tjakopan/yarl/retry/AsyncWaitAndRetryHandleResultTest.java
package hr.tjakopan.yarl.retry;

import hr.tjakopan.yarl.Context;
import hr.tjakopan.yarl.DelegateResult;
import hr.tjakopan.yarl.test.helpers.TestResult;
import org.junit.Test;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import static hr.tjakopan.yarl.Functions.fromConsumer4Async;
import static org.assertj.core.api.Assertions.assertThat;

public class AsyncWaitAndRetryHandleResultTest {
  @Test
  public void shouldBeAbleToCalculateRetryDurationsBasedOnTheHandledFault() {
    final var expectedRetryWaits = Map.of(TestResult.FAULT, Duration.ofMillis(2),
      TestResult.FAULT_AGAIN, Duration.ofMillis(4));
    final var actualRetryWaits = new ArrayList<Duration>(2);
    final var policy = AsyncRetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .waitAndRetry(2,
        (Integer i, DelegateResult<TestResult> outcome, Context c) ->
          outcome.fold(expectedRetryWaits::get, e -> Duration.ZERO),
        fromConsumer4Async(dr -> duration -> (i, c) -> actualRetryWaits.add(duration)));

    final var iterator = expectedRetryWaits.keySet().iterator();
    policy.executeAsync(() -> {
      if (iterator.hasNext()) {
        return CompletableFuture.completedFuture(iterator.next());
      } else {
        return CompletableFuture.completedFuture(TestResult.UNDEFINED);
      }
    })
      .join();

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values());
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/retry/AsyncWaitAndRetryHandleResultTest.kt
package hr.tjakopan.yarl.retry

import hr.tjakopan.yarl.Policy
import hr.tjakopan.yarl.test.helpers.TestResult
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.assertj.core.api.Assertions.assertThat
import java.time.Duration
import kotlin.test.Test

@ExperimentalCoroutinesApi
class AsyncWaitAndRetryHandleResultTest {
  @Test
  fun `should be able to calculate retry durations based on the handled fault`() = runBlockingTest {
    val expectedRetryWaits =
      mapOf(TestResult.FAULT to Duration.ofSeconds(2), TestResult.FAULT_AGAIN to Duration.ofSeconds(4))
    val actualRetryWaits = mutableListOf<Duration>()
    val policy = Policy.asyncRetry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .waitAndRetry(
        2,
        { _, outcome, _ ->
          outcome.fold(
            { expectedRetryWaits[it] },
            { Duration.ZERO }) ?: Duration.ZERO
        }) { _, duration, _, _ -> actualRetryWaits.add(duration) }

    val iterator = expectedRetryWaits.iterator()
    policy.execute {
      if (iterator.hasNext()) {
        return@execute iterator.next().key
      } else {
        return@execute TestResult.UNDEFINED
      }
    }

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values)
  }
}
